### PR TITLE
[Suggestion] Support for multiple projects with deliverable output

### PR DESCRIPTION
**Pull request by Vidarls**
Following up fsprojects/ProjectScaffold#101

Managed to clean up and simplify quite a bit.

### CODE REVIEW

**Comment by tpetricek** [comment]
In a way, I quite like the simplicity of dropping everything in a single `bin` folder - though it has obvious drawbacks.

What do others think about this? /cc @forki @pblasucci @sergey-tihon ?

**Comment by sergey-tihon** [comment]
I also like an approach with single `bin` folder. 

The one drawback that I see is the case when you want to build >1 NuGet package from the repo. In this case you probably want to build binaries to each package to different directories.

**Comment by pblasucci** [comment]
Personally, I'm a fan of the single `bin` directory. But I can see the merits of different paths to support multiple NuGet packages. 

Perhaps something that complicated is beyond the scope of Scaffold? After all, the original intent was to provide a quickly running, high-quality baseline. I always imagined some level of ad hoc tweaking would be done (esp. once you fall outside the typical use cases).

**Comment by Vidarls** [comment]
For my current setup I enjoy some advantages from separate bin folders (particularly CI server download single sub proj output as zip), but I really don't have strong feelings about this anyway.

The only practical issue here is to decide what files to put in what nuget package.
But I don't think it needs to be very difficult using either single bin, or bin/proj approach. Even with multiple packages being output. One can use the src/proj file structure to get the dll names.
(Only becomes a problem if you have some libary type packages where all dependencies are specified in the nuspec, and some deployment packages that are intended to be self contained.)

In my opinion having multiple packages is not really a strange pattern. NancyFx does it, simple.data does it, the new Xunit does it: Core package and addition packages for different feature sets.

The most important thing to me is to have a sensible default that can easily be tweaked. That way I can base my projects on scaffold, modify them, and later keep up to date by pulling in new changes.

**Comment by tpetricek** [comment]
I'm happy to leave the decision to others here - though I'm a bit worried about the increasing complexity of ProjectScaffold... (Not that I would have any good solution to this.)

**Comment by Vidarls** [comment]
Now that Paket supports packing, I suppose quite a bit of the code here is redundant.
Is anyone working on integrating Paket packaging (as opposed to nuget packaging) in ProjectScaffold?

If not, I am entering a infrastructure maintenance period this week, and I have time to have a go.

**Comment by forki** [comment]
I planned to do this, but I'd prefer if you could do it since we need more
people trying it out and find bugs/unclear behavior.

**Comment by Vidarls** [comment]
Ok, then I will have a go.
(I'll probably start by trying to replace the existing single output packing, then move on to the multiproject version).

**Comment by Vidarls** [comment]
I think this one could use a new review.

Not many changes now :)

__For future dev:__

If we could expose the  `Paket.pack` api for finding the output nodes, and feed those to Fsharp.Formatting for generating reference docs, we no longer need a hard-coded bin folder for anything other than convenience of knowing where to look for binaries.

No more need for the copybinaries part of this pull req, and anyone would be free to use any bin path they want (ref above discussion of single bucket or subfolder structure in output bin folder)

**Comment by Vidarls** [comment]
Might there be some Mono / Linux issues with the `CopyDir` function of FAKE?

**Comment by forki** [comment]
> Might there be some Mono / Linux issues with the CopyDir function of FAKE?
I don't know of any.

**Comment by Vidarls** [comment]
You're right, its a case of windows guy running stuff on linux, forgetting __cAsINg__ matters..

**Comment by Vidarls** [comment]
Is this again the API limig thing?

```
Paket version 0.31.1.0
found: /home/travis/build/fsprojects/ProjectScaffold/paket.dependencies
Paket failed with:
	Could not retrieve data from https://api.github.com/repos/fsharp/FAKE/commits/master

Message: The remote server returned an error: (403) Forbidden.

```

**Comment by forki** [comment]
Yep. But seems it's only happening on TravisCI

**Comment by Vidarls** [comment]
Ok, can you go through and give some feedback on this PR when you have the time?

**Comment by Vidarls** [comment]
Any feedback on what is missing before this one can be merged?

File: paket.dependencies [hunk 1]
```diff
@@ -5,5 +5,6 @@ nuget NUnit
 nuget NUnit.Runners
 nuget FAKE
 nuget SourceLink.Fake
+nuget Paket.Core
```
Comment (forki) [inline paket.dependencies]: do we still need this?

File: src/FSharp.ProjectTemplate/FSharp.ProjectTemplate.fsproj [hunk 2]
```diff
@@ -19,19 +19,19 @@
     <DebugType>full</DebugType>
     <Optimize>false</Optimize>
     <Tailcalls>false</Tailcalls>
-    <OutputPath>..\..\bin</OutputPath>
+    <OutputPath>.\bin\Debug</OutputPath>
     <DefineConstants>DEBUG;TRACE</DefineConstants>
     <WarningLevel>3</WarningLevel>
-    <DocumentationFile>..\..\bin\FSharp.ProjectTemplate.xml</DocumentationFile>
+    <DocumentationFile>.\bin\Debug\FSharp.ProjectTemplate.xml</DocumentationFile>
```
Comment (forki) [inline src/FSharp.ProjectTemplate/FSharp.ProjectTemplate.fsproj]: should be FSharp.ProjectScaffold, right?

**Comment by forki** [comment]
can you please rebase and squash this?

File: docs/tools/generate.template [hunk 3]
```diff
@@ -85,20 +83,29 @@ let references =
       Path.GetFullPath "./../../packages/FSharp.Formatting/lib/net40/FSharp.MetadataFormat.dll" ]
     |> Some
   else None
+let findBinaries () =
```
Comment (forki) [inline docs/tools/generate.template]: how about making this a simple value. Like `let binaries =`

File: paket.dependencies (hunk 1 above)
Comment (Vidarls) [inline paket.dependencies]: nope, It's all in the Fake helper, so this one can go

File: src/FSharp.ProjectTemplate/FSharp.ProjectTemplate.fsproj (hunk 2 above)
Comment (Vidarls) [inline src/FSharp.ProjectTemplate/FSharp.ProjectTemplate.fsproj]: ?
The project is named `FSharp.ProjectTemplate` so I assume the xml docs should have the same name. (Unless I'm completely misunderstanding you question here)

File: docs/tools/generate.template (hunk 3 above)
Comment (Vidarls) [inline docs/tools/generate.template]: sounds good

File: src/FSharp.ProjectTemplate/FSharp.ProjectTemplate.fsproj (hunk 2 above)
Comment (forki) [inline src/FSharp.ProjectTemplate/FSharp.ProjectTemplate.fsproj]: ok. my question is more: this gets correctly replaced, right?

**Comment by Vidarls** [comment]
Regarding rebasing and squashing: Will rebase and squash break my own downstream projects (already based on the code in this PR) once this is merged and I pull changes from master

**Comment by forki** [comment]
ok then forget the rebase ;-)

File: src/FSharp.ProjectTemplate/FSharp.ProjectTemplate.fsproj (hunk 2 above)
Comment (Vidarls) [inline src/FSharp.ProjectTemplate/FSharp.ProjectTemplate.fsproj]: yes,

from: https://github.com/fsprojects/ProjectScaffold/blob/master/init.fsx#L138

```f#
 |> replace projectTemplateName projectName

```

**Comment by Vidarls** [comment]
forgot a question mark there, it was genuinely meant as a question, as to what will happen to my downstream projects if I rebase? I don't know..

File: src/FSharp.ProjectTemplate/FSharp.ProjectTemplate.fsproj (hunk 2 above)
Comment (forki) [inline src/FSharp.ProjectTemplate/FSharp.ProjectTemplate.fsproj]: :+1:

**Comment by forki** [comment]
no. It's fine. please just correct the last two minor things. then I'll merge it.

**Comment by Vidarls** [comment]
Minor stuff fixed. Don't understand quite why the travis build fails.

**Comment by forki** [comment]
Looks like the directory doesn't exist

**Comment by Vidarls** [comment]
tries to find this directory:

`/home/travis/build/fsprojects/ProjectScaffold/src/home/bin/Release`

Notise the __home__ between _src_ and _bin_

File: build.template [hunk 4]
```diff
@@ -100,6 +99,16 @@ Target "AssemblyInfo" (fun _ ->
   csProjs |> Seq.iter genCSAssemblyInfo
 )
 
+// Copies binaries from default VS location to exepcted bin folder
+// But keeps a subdirectory structure for each project in the 
+// src folder to support multiple project outputs
+Target "CopyBinaries" (fun _ ->
+    directoryInfo "src" 
+    |> subDirectories 
+    |> Array.map (fun d -> (d.Name,  sprintf "%s/bin/Release" d.FullName))
```
Comment (forki) [inline build.template]: Please use Path.Combine

File: build.template [hunk 5]
```diff
@@ -100,6 +99,16 @@ Target "AssemblyInfo" (fun _ ->
...
 )
 
+// Copies binaries from default VS location to exepcted bin folder
+// But keeps a subdirectory structure for each project in the 
+// src folder to support multiple project outputs
+Target "CopyBinaries" (fun _ ->
+    directoryInfo "src" 
+    |> subDirectories 
+    |> Array.map (fun d -> (d.Name,  sprintf "%s/bin/Release" d.FullName)) 
+    |> Array.iter (fun (n, p) -> CopyDir (sprintf "bin/%s" n) p (fun _ -> true))
```
Comment (forki) [inline build.template]: Same here please use path.combine or @@

**Comment by Vidarls** [comment]
weird.

It seems 

```f#
directoryInfo "src" 
    |> subDirectories 
```

Is returning an extra sub directory:

```
Starting Target: CopyBinaries (==> Build)
Creating /home/travis/build/fsprojects/ProjectScaffold/bin/CONTINUOUSINTEGRATION
Creating /home/travis/build/fsprojects/ProjectScaffold/bin/home
Running build failed.

Error:
System.IO.DirectoryNotFoundException: Directory '/home/travis/build/fsprojects/ProjectScaffold/src/home/bin/Release' not found.
```

**Comment by forki** [comment]
wow. sounds like an issue in FAKE. can you explicitly exlude it for now?

**Comment by Vidarls** [comment]
Finally :tada:

**Comment by forki** [comment]
Good work. Thanks

**Pull request merged**